File: src/backtracing/_22_generate_parenthesis.rs
struct Solution;

impl Solution {
    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        fn helper(left: i32, right: i32, cur: &mut String, ret: &mut Vec<String>) {
            if left == 0 && right == 0 {
                ret.push(cur.clone());
                return;
            }
            if left > right {
                return;
            }
            if left > 0 {
                cur.push_str("(");
                helper(left - 1, right, cur, ret);
                cur.pop();
            }

            if right > 0 {
                cur.push_str(")");
                helper(left, right - 1, cur, ret);
                cur.pop();
            }
        }
        if n == 0 {
            return vec![];
        }
        let mut ret = vec![];
        helper(n, n, &mut String::new(), &mut ret);
        ret
    }
}

File: hot-100/22_generateParenthesis.go
package hot_100

func generateParenthesis(n int) []string {
	ans := []string{}

	cur := make([]byte, 0)
	var dfs func(left, right int)
	dfs = func(left, right int) {
		if left == 0 && right == 0 {
			tmp := make([]byte, len(cur))
			copy(tmp, cur)
			ans = append(ans, string(cur))
			return
		}
		if left > right {
			return
		}
		if left > 0 {
			cur = append(cur, '(')
			dfs(left-1, right)
			cur = cur[:len(cur)-1]
		}
		if right > 0 {
			cur = append(cur, ')')
			dfs(left, right-1)
			cur = cur[:len(cur)-1]
		}
	}
	dfs(n, n)
	return ans
}
